components/style/shared_lock.rs

```diff
@@ -0,0 +1,183 @@
+pub struct SharedRwLock {
+    #[cfg_attr(feature = "servo", ignore_heap_size_of = "Arc")]
+    arc: Arc<RwLock<()>>,
+}
+
+impl fmt::Debug for SharedRwLock {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("SharedRwLock")
+    }
+}
+
+impl SharedRwLock {
+    /// Create a new shared lock
+    pub fn new() -> Self {
+        SharedRwLock {
+            arc: Arc::new(RwLock::new(()))
+        }
+    }
+
+    /// Wrap the given data to make its access protected by this lock.
+    pub fn wrap<T>(&self, data: T) -> Locked<T> {
+        Locked {
+            shared_lock: self.clone(),
+            data: UnsafeCell::new(data),
+        }
+    }
+
+    /// Obtain the lock for reading
+    pub fn read(&self) -> SharedRwLockReadGuard {
+        self.arc.raw_read();
+        SharedRwLockReadGuard {
+            shared_lock: self
+        }
+    }
+
+    /// Obtain the lock for writing
+    pub fn write(&self) -> SharedRwLockWriteGuard {
+        self.arc.raw_write();
+        SharedRwLockWriteGuard {
+            shared_lock: self
+        }
+    }
+
+    fn same_as(&self, other: &Self) -> bool {
+        // FIXME: Use Arc::ptr_eq once it’s stable. https://github.com/rust-lang/rust/issues/36497
+        let a: *const RwLock<()> = Arc::deref(&self.arc);
+        let b: *const RwLock<()> = Arc::deref(&other.arc);
+        a == b
+    }
+}
+
+/// Data protect by a shared lock.
+pub struct Locked<T> {
+    shared_lock: SharedRwLock,
+    data: UnsafeCell<T>,
+}
+
+// Unsafe: the data inside `UnsafeCell` is only accessed in `read_with` and `write_with`,
+// where guards ensure synchronization.
+unsafe impl<T: Send> Send for Locked<T> {}
+unsafe impl<T: Send + Sync> Sync for Locked<T> {}
+
+impl<T: fmt::Debug> fmt::Debug for Locked<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let guard = self.shared_lock.read();
+        self.read_with(&guard).fmt(f)
+    }
+}
+
+impl<T> Locked<T> {
+    /// Access the data for reading.
+    pub fn read_with<'a>(&'a self, guard: &'a SharedRwLockReadGuard) -> &'a T {
+        assert!(self.shared_lock.same_as(&guard.shared_lock),
+                "StyleRef::read_with called with a guard from an unrelated SharedRwLock");
+        let ptr = self.data.get();
+
+        // Unsafe:
+        //
+        // * The guard guarantees that the lock is taken for reading,
+        //   and we’ve checked that it’s the correct lock.
+        // * The returned reference borrows *both* the data and the guard,
+        //   so that it can outlive neither.
+        unsafe {
+            &*ptr
+        }
+    }
+
+    /// Access the data for writing.
+    pub fn write_with<'a>(&'a self, guard: &'a mut SharedRwLockWriteGuard) -> &'a mut T {
+        assert!(self.shared_lock.same_as(&guard.shared_lock),
+                "StyleRef::write_with called with a guard from an unrelated SharedRwLock");
+        let ptr = self.data.get();
+
+        // Unsafe:
```

Comment:
In `fn write_with<'a>(&'a self, guard: &'a mut SharedRwLockWriteGuard) -> &'a mut T` the return value borrows *both* `self` and `guard` (since they use the same lifetime parameter).